csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -362,18 +362,46 @@ TensorView* maybeDoReplacement(TensorView* orig) {
       replacement->toString(),
       " has different dimension than original ",
       orig->toString());
+  // Replacing `orig` with `replacement` can introduce resharding on the
+  // consumer of `orig`, e.g., the `cast` in the following example.
+  //
+  // [bDID{1}, i0]: replacement
+  //      |
+  //      | squeeze
+  //      v
+  //     [i0]
+  //      |
+  //      | broadcast
+  //      v
+  //  [b{1}, i0]: orig
+  //      |
+  //      | cast
+  //      v
+  //  [b{1}, i0]
+  //
+  // Such resharding expressions won't be resolved by `insert_reshardings`
+  // because `insert_reshardings` runs before `remove_bcast_squeeze`.
+  // Therefore, if resharding is needed, instead of replacing `orig` with
+  // `replacement`, we link them with a resharding `set`.
+  bool needs_resharding = false;
   for (size_t i : c10::irange(old_loop.size())) {
-    if (old_loop[i]->isParallelized()) {
+    if (old_loop[i]->getParallelType() != new_loop[i]->getParallelType()) {
       NVF_ERROR(
-          old_loop[i]->isDeviceDim(),
+          old_loop[i]->isDeviceDim() || new_loop[i]->isDeviceDim(),
           "Before scheduling, we expect the only parallelized ",
           "dimensions to be device dims");
-      // In particular, we might have a Device dimension parallelized for the
-      // output, which we need to preserve.
-      new_loop[i]->parallelize(old_loop[i]->getParallelType());
+      needs_resharding = true;
+      break;
     }
   }
-  return replacement;
+
+  if (needs_resharding) {
```

Comment:
I'm also fine with simply bailing out and thus keeping the squeeze-broadcast pair. I send this version because it's slightly more powerful.

Reply:
This makes sense to me. If we bail then we would leave the squeeze+bcast sequence shown in your comment above, which could lead to some of the errors we have seen previously, so I prefer to leave this as you've written it here.